src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -236,64 +237,170 @@ object DeployPlugin extends AutoPlugin {
+    val deploys: Iterable[Future[Try[Unit]]] = hosts flatMap { hostConfig =>
+      val deployHost = hostConfig.host
+      val baseDeployDir = deployConfig.baseDirectory
 
-          Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
-        }
-        val deployHost = hostConfig.host
-        val deployDirectory = hostConfig.directory
-
-        val rsyncDirs = deployDirs.value map (name => s"--include=/$name")
-        val rsyncCommand = Seq.concat(
-          Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")),
-          rsyncDirs,
-          Seq(
-            "--exclude=/*",
-            "--delete",
-            universalStagingDir.getPath + "/",
-            deployHost + ":" + deployDirectory
-          )
-        )
+      // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh operations.
+      val sshCommand = {
+        val keyFile = sys.env.getOrElse("AWS_PEM_FILE", {
+          throw new IllegalStateException(s"Environment variable 'AWS_PEM_FILE' is undefined")
+        })
+        require(keyFile.nonEmpty, "Environment variable 'AWS_PEM_FILE' is empty")
 
-        // Shell-friendly version of rsync command, with rsh value quoted.
-        val quotedRsync = rsyncCommand.patch(
-          2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
-        ).mkString(" ")
-
-        // Now, ssh to the remote host and run the restart script.
-        val restartScript = deployDirectory + "/" + hostConfig.startupScript
-        val restartArgs = if (hostConfig.startupArgs.nonEmpty) {
-          "--" +: hostConfig.startupArgs
-        } else {
-          Seq()
-        }
-
-        val restartCommand = Seq.concat(
-          sshCommand,
-          Seq(deployHost, restartScript, "restart"),
-          restartArgs
-        )
+        Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
+      }
 
-        // Run commands within a Try so we can collect errors using Future.fold.
-        // TODO(danm): Use ProcessIO to capture and report the actual stderr of failures here.
-        Try {
-          log.info("Running " + quotedRsync + " . . .")
-          if (Process(rsyncCommand).! != 0) {
-            sys.error(s"Error running rsync to host '$deployHost'.")
-          }
+      val replicas = deployConfig.replicaOverridesByHost(hostConfig)
+      val numReplicas = replicas.length
+
+      // First, stop and remove any replicas on the remote that won't be restarted
+      // as part of the deploy.
+      val (deployParent, deployTarget) = baseDeployDir.splitAt(baseDeployDir.lastIndexOf('/'))
+      // Strip leading / from deployTarget so we can match names on it.
+      val namePattern = deployTarget.tail
+      // Build a regex matching the replicas we will eventually update / restart as part of this
+      // process, and therefore DON'T want to stop and remove.
+      val restartedPattern = if (numReplicas == 1) {
+        namePattern
+      } else {
+        s"$namePattern-[1-$numReplicas]"
```

Comment:
Some of the complexity of this block comes from the fact that we don't want to include a replica number for deploys pointing at only one host.  My main worry was the case when you jump from pointing at one remote to pointing at > 1 remote.

For example, we're running one `inference-solver` in dev right now.  On the host machine there's a directory `/local/deploy/inference-solver`, and we don't want to change that unless we add a second instance.  When we do add a second instance, the directory names will become `/local/deploy/inference-solver-1.sh` and `/local/deploy/inference-solver-2.sh`.  Presumably the config for the first solver (which already existed) won't change, which means the solver in `inference-solver-1` will be trying to bind to the same port as the solver running in `inference-solver`.  The existing restart commands won't handle stopping the solver in `inference-solver` because they only run in the newly-deployed directories, and the deploy will fail when `inference-solver-1` fails to bind to its port.